helix-core/src/history.rs

```diff
@@ -23,72 +64,259 @@ impl Default for History {
+            Some(instant) => self.jump_instant(instant),
+            None => self.jump_to(0),
+        }
+    }
+
+    fn jump_duration_forward(&mut self, duration: Duration) -> Vec<Transaction> {
+        match self.revisions[self.current].timestamp.checked_add(duration) {
+            Some(instant) => self.jump_instant(instant),
+            None => self.jump_to(self.revisions.len() - 1),
+        }
+    }
+
+    pub fn earlier(&mut self, uk: UndoKind) -> Vec<Transaction> {
+        use UndoKind::*;
+        match uk {
+            Steps(n) => self.jump_backward(n),
+            TimePeriod(d) => self.jump_duration_backward(d),
+        }
+    }
+
+    pub fn later(&mut self, uk: UndoKind) -> Vec<Transaction> {
+        use UndoKind::*;
+        match uk {
+            Steps(n) => self.jump_forward(n),
+            TimePeriod(d) => self.jump_duration_forward(d),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum UndoKind {
+    Steps(usize),
+    TimePeriod(std::time::Duration),
+}
+
+const TIME_UNITS: &[&str] = &["seconds", "minutes", "hours", "days"];
+
+fn find_time_unit(s: &str) -> Result<&'static str, String> {
+    for unit in TIME_UNITS {
+        if unit.starts_with(&s.to_lowercase()) {
+            return Ok(unit);
+        }
+    }
+    Err(format!("incorrect time unit: {}", s))
+}
+
+static DURATION_VALIDATION_REGEX: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"^(?:\d+\s*[a-zA-Z]+\s*)+$").unwrap());
+
+static NUMBER_UNIT_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"(\d+)\s*([a-zA-Z]+)").unwrap());
+
+fn parse_human_duration(s: &str) -> Result<Duration, String> {
+    if !DURATION_VALIDATION_REGEX.is_match(s) {
+        return Err("duration should be composed \
+        of positive integers followed by time units"
+            .to_string());
+    }
+
+    let mut units = std::collections::HashMap::new();
+    for cap in NUMBER_UNIT_REGEX.captures_iter(s) {
+        let (n_str, short_unit) = (&cap[1], &cap[2]);
+
+        let n = match n_str.parse::<u64>() {
+            Err(_) => return Err(format!("integer too large: {}", n_str)),
+            Ok(n) => n,
+        };
+
+        let unit = find_time_unit(short_unit)?;
+        if units.contains_key(unit) {
+            return Err(format!("{} specified more than once", unit));
+        }
+        units.insert(unit, n);
+    }
 
-        // for now, simply pick the latest child (linear undo / redo)
-        if let Some((index, transaction)) = current_revision.children.last() {
-            self.cursor = *index;
+    let units_in_seconds = [
+        Some(*units.get("seconds").unwrap_or(&0)),
+        units.get("minutes").unwrap_or(&0).checked_mul(60),
+        units.get("hours").unwrap_or(&0).checked_mul(60 * 60),
+        units.get("days").unwrap_or(&0).checked_mul(60 * 60 * 24),
+    ];
+    // A checked sum.
+    let seconds = units_in_seconds
+        .iter()
+        .try_fold(0u64, |acc, x| x.and_then(|x| acc.checked_add(x)));
```

Comment:
I wonder if we should use `saturating_add` but nevermind an error here can give less surprises. If you follow what I mentioned above.
```
    let [s, m, h, d] = units;
    let (m, o1) = m.overflowing_mul(60);
    let (h, o2) = h.overflowing_mul(60 * 60);
    let (d, o3) = d.overflowing_mul(24 * 60 * 60);
    let (s1, o4) = s.overflowing_add(m);
    let (s2, o5) = h.overflowing_add(d);
    let s = if o1 | o2 | o3 | o4 | o5 {
        None
    } else {
        s1.checked_add(s2)
    };
```
Will this be better?

Reply:
I don't feel like my version is very pretty and I'm not attached to it, but I don't feel your proposal is an improvement. It's longer and you need to manually name and type out the booleans that signify overflow. There might be a more elegant way to write this, but I don't think there's a lot to gain by searching for it.